File: frontend/src/types/incident.ts
/**
 * Incident domain types. Catalog values are stored in English (they match the
 * backend `StrEnum`s exactly); the UI translates them only for display.
 */

export const INCIDENT_SEVERITIES = ["Critical", "High", "Medium", "Low"] as const;
export type IncidentSeverity = (typeof INCIDENT_SEVERITIES)[number];

export const INCIDENT_STATUSES = [
  "Open",
  "Investigating",
  "Identified",
  "Monitoring",
  "Resolved",
  "Closed",
] as const;
export type IncidentStatus = (typeof INCIDENT_STATUSES)[number];

/** Non-terminal statuses - an incident in one of these is "open". */
export const ACTIVE_INCIDENT_STATUSES: readonly IncidentStatus[] = [
  "Open",
  "Investigating",
  "Identified",
  "Monitoring",
];
export const TERMINAL_INCIDENT_STATUSES: readonly IncidentStatus[] = ["Resolved", "Closed"];

export const INCIDENT_PRIORITIES = ["P1", "P2", "P3", "P4"] as const;
export type IncidentPriority = (typeof INCIDENT_PRIORITIES)[number];

export const INCIDENT_EVENT_TYPES = [
  "CREATED",
  "STATUS_CHANGED",
  "SEVERITY_CHANGED",
  "PRIORITY_CHANGED",
  "OWNER_CHANGED",
  "ASSET_ADDED",
  "ASSET_REMOVED",
  "COMMENT",
  "RESOLVED",
  "REOPENED",
] as const;
export type IncidentEventType = (typeof INCIDENT_EVENT_TYPES)[number];

/** A row in the incidents list. */
export interface Incident {
  id: string;
  title: string;
  severity: IncidentSeverity;
  status: IncidentStatus;
  priority: IncidentPriority;
  owner: string | null;
  started_at: string;
  detected_at: string | null;
  resolved_at: string | null;
  affected_asset_count: number;
  created_at: string;
  updated_at: string;
}

/** A compact projection of an affected asset, embedded in the incident detail. */
export interface IncidentAssetRef {
  id: string;
  name: string;
  asset_type: string;
  environment: string;
  criticality: string;
  status: string;
  is_active: boolean;
  /** Non-null when the asset has been moved to Trash - the relationship is kept
   *  (history) and the UI badges it "En papelera". */
  deleted_at: string | null;
}

export interface IncidentEvent {
  id: string;
  type: IncidentEventType;
  message: string;
  created_by: string | null;
  actor_email: string | null;
  created_at: string;
}

/** Full incident detail: metadata + affected assets + timeline. */
export interface IncidentDetail {
  id: string;
  title: string;
  description: string | null;
  severity: IncidentSeverity;
  status: IncidentStatus;
  priority: IncidentPriority;
  owner: string | null;
  started_at: string;
  detected_at: string | null;
  resolved_at: string | null;
  created_by: string;
  created_at: string;
  updated_at: string;
  affected_assets: IncidentAssetRef[];
  timeline: IncidentEvent[];
}

export interface IncidentPage {
  items: Incident[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

/** Aggregate incident counts from `GET /api/v1/incidents/summary`. */
export interface IncidentSummary {
  total: number;
  open: number;
  critical_open: number;
  investigating: number;
  monitoring: number;
  resolved_recently: number;
  by_severity: Record<string, number>;
  by_status: Record<string, number>;
}

/** Fields the create form submits. */
export interface IncidentCreateInput {
  title: string;
  description?: string | null;
  severity: IncidentSeverity;
  priority: IncidentPriority;
  status?: IncidentStatus;
  owner?: string | null;
  started_at?: string | null;
  detected_at?: string | null;
  asset_ids?: string[];
}

/** Partial update - any subset of the editable fields. */
export type IncidentUpdateInput = Partial<IncidentCreateInput>;

function isOneOf<T extends readonly string[]>(list: T, v: unknown): v is T[number] {
  return typeof v === "string" && (list as readonly string[]).includes(v);
}

function isCountMap(v: unknown): v is Record<string, number> {
  return (
    typeof v === "object" &&
    v !== null &&
    Object.values(v as Record<string, unknown>).every((n) => typeof n === "number")
  );
}

export function isIncident(value: unknown): value is Incident {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  return (
    typeof v.id === "string" &&
    typeof v.title === "string" &&
    isOneOf(INCIDENT_SEVERITIES, v.severity) &&
    isOneOf(INCIDENT_STATUSES, v.status) &&
    isOneOf(INCIDENT_PRIORITIES, v.priority) &&
    (v.owner === null || typeof v.owner === "string") &&
    typeof v.started_at === "string" &&
    (v.detected_at === null || typeof v.detected_at === "string") &&
    (v.resolved_at === null || typeof v.resolved_at === "string") &&
    typeof v.affected_asset_count === "number" &&
    typeof v.created_at === "string" &&
    typeof v.updated_at === "string"
  );
}

export function isIncidentPage(value: unknown): value is IncidentPage {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  return (
    Array.isArray(v.items) &&
    v.items.every(isIncident) &&
    typeof v.page === "number" &&
    typeof v.page_size === "number" &&
    typeof v.total === "number" &&
    typeof v.total_pages === "number"
  );
}

function isEvent(value: unknown): value is IncidentEvent {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  return (
    typeof v.id === "string" &&
    isOneOf(INCIDENT_EVENT_TYPES, v.type) &&
    typeof v.message === "string" &&
    (v.created_by === null || typeof v.created_by === "string") &&
    (v.actor_email === null || typeof v.actor_email === "string") &&
    typeof v.created_at === "string"
  );
}

export function isIncidentDetail(value: unknown): value is IncidentDetail {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  return (
    typeof v.id === "string" &&
    typeof v.title === "string" &&
    (v.description === null || typeof v.description === "string") &&
    isOneOf(INCIDENT_SEVERITIES, v.severity) &&
    isOneOf(INCIDENT_STATUSES, v.status) &&
    isOneOf(INCIDENT_PRIORITIES, v.priority) &&
    typeof v.created_by === "string" &&
    Array.isArray(v.affected_assets) &&
    Array.isArray(v.timeline) &&
    v.timeline.every(isEvent)
  );
}

export function isIncidentSummary(value: unknown): value is IncidentSummary {
  if (typeof value !== "object" || value === null) return false;
  const v = value as Record<string, unknown>;
  return (
    typeof v.total === "number" &&
    typeof v.open === "number" &&
    typeof v.critical_open === "number" &&
    typeof v.investigating === "number" &&
    typeof v.monitoring === "number" &&
    typeof v.resolved_recently === "number" &&
    isCountMap(v.by_severity) &&
    isCountMap(v.by_status)
  );
}

File: backend/app/schemas/incident.py
"""Request/response schemas for the Incidents API.

Catalog fields reuse the ``StrEnum`` types from the ORM model, so Pydantic
rejects any value outside the vocabulary with a ``422``. ``IncidentCreate`` /
``IncidentUpdate`` set ``extra="forbid"`` so an unknown field is a client error.

Fields the client must never set are simply absent from the input models:

* ``created_by``  - derived from the authenticated user in the route.
* ``resolved_at`` - derived from ``status`` transitions by the service layer.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Annotated

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.models.asset import AssetStatus, AssetType, Criticality, Environment
from app.models.incident import (
    DESCRIPTION_MAX_LENGTH,
    EVENT_MESSAGE_MAX_LENGTH,
    OWNER_MAX_LENGTH,
    TITLE_MAX_LENGTH,
    IncidentEventType,
    IncidentPriority,
    IncidentSeverity,
    IncidentStatus,
)
from app.schemas.asset import MessageResponse

__all__ = [
    "DEFAULT_PAGE_SIZE",
    "MAX_ASSET_LINKS",
    "MAX_PAGE_SIZE",
    "IncidentAssetRef",
    "IncidentCreate",
    "IncidentEventRead",
    "IncidentListItem",
    "IncidentPage",
    "IncidentRead",
    "IncidentSummary",
    "IncidentUpdate",
    "MessageResponse",
]

MAX_PAGE_SIZE = 100
# Incident rows are denser than asset rows (title + severity + status + priority
# + affected count + owner + two dates); 15 keeps the first page scannable.
DEFAULT_PAGE_SIZE = 15
# Upper bound on affected assets linked in a single request - guards the
# association writes and the ASSET_ADDED/ASSET_REMOVED event fan-out.
MAX_ASSET_LINKS = 200

TitleField = Annotated[str, Field(min_length=1, max_length=TITLE_MAX_LENGTH)]
OwnerField = Annotated[str | None, Field(default=None, max_length=OWNER_MAX_LENGTH)]
DescriptionField = Annotated[
    str | None, Field(default=None, max_length=DESCRIPTION_MAX_LENGTH)
]
AssetIdsField = Annotated[
    list[uuid.UUID], Field(default_factory=list, max_length=MAX_ASSET_LINKS)
]
CommentField = Annotated[str, Field(min_length=1, max_length=EVENT_MESSAGE_MAX_LENGTH)]


def _clean_optional(value: str | None) -> str | None:
    if value is None:
        return None
    trimmed = value.strip()
    return trimmed or None


def _dedupe(ids: list[uuid.UUID]) -> list[uuid.UUID]:
    seen: set[uuid.UUID] = set()
    out: list[uuid.UUID] = []
    for i in ids:
        if i not in seen:
            seen.add(i)
            out.append(i)
    return out


class IncidentCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: TitleField
    description: DescriptionField
    severity: IncidentSeverity
    priority: IncidentPriority
    status: IncidentStatus = IncidentStatus.OPEN
    owner: OwnerField
    started_at: datetime | None = None
    detected_at: datetime | None = None
    asset_ids: AssetIdsField

    @field_validator("title")
    @classmethod
    def _strip_title(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("must not be blank")
        return stripped

    @field_validator("owner", "description")
    @classmethod
    def _strip_optional(cls, v: str | None) -> str | None:
        return _clean_optional(v)

    @field_validator("asset_ids")
    @classmethod
    def _dedupe_assets(cls, v: list[uuid.UUID]) -> list[uuid.UUID]:
        return _dedupe(v)


class IncidentUpdate(BaseModel):
    """Partial update. Every field is optional; unknown fields are rejected.

    ``asset_ids`` semantics: **omitted** -> affected assets untouched; provided
    (even ``[]``) -> the affected-asset set is replaced with exactly that list,
    emitting ASSET_ADDED / ASSET_REMOVED timeline events for the difference.
    """

    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=TITLE_MAX_LENGTH)
    description: DescriptionField
    severity: IncidentSeverity | None = None
    priority: IncidentPriority | None = None
    status: IncidentStatus | None = None
    owner: OwnerField
    started_at: datetime | None = None
    detected_at: datetime | None = None
    asset_ids: list[uuid.UUID] | None = Field(default=None, max_length=MAX_ASSET_LINKS)

    @field_validator("title")
    @classmethod
    def _strip_title(cls, v: str | None) -> str | None:
        if v is None:
            return None
        stripped = v.strip()
        if not stripped:
            raise ValueError("must not be blank")
        return stripped

    @field_validator("owner", "description")
    @classmethod
    def _strip_optional(cls, v: str | None) -> str | None:
        return _clean_optional(v)

    @field_validator("asset_ids")
    @classmethod
    def _dedupe_assets(cls, v: list[uuid.UUID] | None) -> list[uuid.UUID] | None:
        return None if v is None else _dedupe(v)


class IncidentCommentCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    message: CommentField

    @field_validator("message")
    @classmethod
    def _strip(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("must not be blank")
        return stripped


class IncidentAssetRef(BaseModel):
    """Compact projection of an affected asset, embedded in the incident detail.

    ``deleted_at`` is non-null when the asset has been moved to Trash: the
    persisted relationship is kept (history), and the UI badges it "En papelera"
    rather than dropping it.
    """

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    asset_type: AssetType
    environment: Environment
    criticality: Criticality
    status: AssetStatus
    is_active: bool
    deleted_at: datetime | None = None


class IncidentEventRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    type: IncidentEventType
    message: str
    created_by: uuid.UUID | None
    actor_email: str | None = None
    created_at: datetime


class IncidentListItem(BaseModel):
    """A row in the incidents list - no description, no timeline."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    severity: IncidentSeverity
    status: IncidentStatus
    priority: IncidentPriority
    owner: str | None
    started_at: datetime
    detected_at: datetime | None
    resolved_at: datetime | None
    affected_asset_count: int = Field(ge=0)
    created_at: datetime
    updated_at: datetime


class IncidentRead(BaseModel):
    """Full incident detail: metadata + affected assets + timeline."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str
    description: str | None
    severity: IncidentSeverity
    status: IncidentStatus
    priority: IncidentPriority
    owner: str | None
    started_at: datetime
    detected_at: datetime | None
    resolved_at: datetime | None
    created_by: uuid.UUID
    created_at: datetime
    updated_at: datetime
    affected_assets: list[IncidentAssetRef]
    timeline: list[IncidentEventRead]


class IncidentPage(BaseModel):
    items: list[IncidentListItem]
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=MAX_PAGE_SIZE)
    total: int = Field(ge=0)
    total_pages: int = Field(ge=0)


class IncidentSummary(BaseModel):
    """Aggregate incident counts for the Dashboard. Read-only, derived from the
    ``incidents`` table. Every catalog key is present even when its count is zero.
    """

    total: int = Field(ge=0)
    open: int = Field(ge=0)
    critical_open: int = Field(ge=0)
    investigating: int = Field(ge=0)
    monitoring: int = Field(ge=0)
    resolved_recently: int = Field(ge=0)
    by_severity: dict[str, int]
    by_status: dict[str, int]
